analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -35,8 +35,22 @@ public sealed record SymbolicValue
         public static readonly SymbolicValue True = NotNull.WithConstraint(BoolConstraint.True);
         public static readonly SymbolicValue False = NotNull.WithConstraint(BoolConstraint.False);
 
+        private readonly ImmutableDictionary<Type, SymbolicConstraint> constraints = ImmutableDictionary<Type, SymbolicConstraint>.Empty;
+        private int? hashCode;
+
         // SymbolicValue can have only one constraint instance of specific type at a time
-        private ImmutableDictionary<Type, SymbolicConstraint> Constraints { get; init; } = ImmutableDictionary<Type, SymbolicConstraint>.Empty;
+        private ImmutableDictionary<Type, SymbolicConstraint> Constraints
+        {
+            get => constraints;
+            init
+            {
+                if (constraints != value)
```

Comment:
This condition is not covered when values are the same. We don't use it that way. And we actually shouldn't, as we should attempt no-operation when possible.

Do we really need the condition then?

Reply:
It doesn't hurt to keep it, IMO. It allows doing something like, e.g. `with Constraints = someCond ? Empty : Constraints`. I could add a test case to increase test coverage if that is a concern.

Reply:
I would remove the condition. It's one more additional `Equals` check on a hot path that doesn't need to be there. 

And so far, we didn't need it. And it's against the intended usage of the property. We can always add it back if we need to.

Reply:
That's fine with me.